File: go-backend-interview-prep/05_errors_and_testing/08_table_driven_tests/main.go
package main

import (
	"fmt"
	"strconv"
)

// ParseIntSafe parses s as base-10 int. Returns fallback on failure.
func ParseIntSafe(s string, fallback int) (int, error) {
	n, err := strconv.Atoi(s)
	if err != nil {
		return fallback, fmt.Errorf("ParseIntSafe(%q): %w", s, err)
	}
	return n, nil
}

// Clamp restricts n to the range [lo, hi].
func Clamp(n, lo, hi int) int {
	if n < lo {
		return lo
	}
	if n > hi {
		return hi
	}
	return n
}

func main() {
	// ParseIntSafe examples
	v, err := ParseIntSafe("42", 0)
	fmt.Printf("ParseIntSafe(\"42\", 0) = %d, err=%v\n", v, err)

	v, err = ParseIntSafe("abc", -1)
	fmt.Printf("ParseIntSafe(\"abc\", -1) = %d, err=%v\n", v, err)

	// Clamp examples
	fmt.Println("Clamp(5, 0, 10) =", Clamp(5, 0, 10))
	fmt.Println("Clamp(-3, 0, 10) =", Clamp(-3, 0, 10))
	fmt.Println("Clamp(15, 0, 10) =", Clamp(15, 0, 10))
}

File: go-backend-interview-prep/05_errors_and_testing/08_table_driven_tests/main.py
"""Table-driven tests -- Python equivalent of the Go example."""


def parse_int_safe(s: str, fallback: int = 0) -> tuple[int, str | None]:
    """Parse s as int. Returns (value, error_msg). Uses fallback on failure."""
    try:
        return int(s), None
    except ValueError:
        return fallback, f"parse_int_safe({s!r}): invalid literal"


def clamp(n: int, lo: int, hi: int) -> int:
    """Restrict n to the range [lo, hi]."""
    if n < lo:
        return lo
    if n > hi:
        return hi
    return n


def main() -> None:
    # ParseIntSafe examples
    v, err = parse_int_safe("42", 0)
    print(f'parse_int_safe("42", 0) = {v}, err={err}')

    v, err = parse_int_safe("abc", -1)
    print(f'parse_int_safe("abc", -1) = {v}, err={err}')

    # Clamp examples
    print(f"clamp(5, 0, 10) = {clamp(5, 0, 10)}")
    print(f"clamp(-3, 0, 10) = {clamp(-3, 0, 10)}")
    print(f"clamp(15, 0, 10) = {clamp(15, 0, 10)}")


if __name__ == "__main__":
    main()
